### PR TITLE
plugins/rust_plugin: rewrite the plugin to improve UX

### PR DESCRIPTION
**Pull request by liushuyu**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

The new plugin will install the Rustup toolchain by default (opt-out possible) and also makes it more intelligent.

Three new options have been added:

- One option for enabling LTO (more optimized binary)
- One option for specifying Rust toolchain channel/version
- One option for disabling default features in the Rust crate

### CODE REVIEW

**Comment by liushuyu** [comment]
The CI errors are unrelated (due to new behaviors in the newer version of `pyright`)

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [inline craft_parts/plugins/rust_plugin.py:165]:
```suggestion
            config_cmd.extend(["--features", *options.rust_features])
```

**Review by lengau — DISMISSED** [review]
Thanks for this! It changes the default behaviour in a way that could have surprising results for existing projects, so I've made suggestions about how to handle backwards compatibility. I'm a big fan of the `rustup` integration though

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:130]: > I'll defer to @cmatsuoka and @mr-cal on this as the maintainer and author, but we may want to default to getting rust from the system package manager.

That's a bad idea. Our in-archive Rust toolchain is usually lagging, and Rust is a very fast-moving language that releases a new compiler version every six weeks.



> To keep backwards compatibility, I would guess the logic could be something like:
> 
>     1. If a `rust-deps` part dependency is defined, no-op (maybe error if a `rust-channel` is defined).
> 
>     2. If a `rust-channel` is defined, use rustup.
> 
>     3. If neither of those, include `rustc` and `cargo` in the build packages.

I chose the current behavior because if you do a quick search on the Snapcraft forum, you will find most users are confused by this (they are migrating from `core18`, which they would expect the plugin to install a `rustup` binary.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:165]: This suggested change does not seem correct.
`cargo` expect all the features to be a string:

`cargo build --features 'a b c'`

instead of:

`cargo build --features a b c`

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:155]: That's a good point. However, I do think if the user chose the update channel instead of a fixed version, they should expect the version to be updated each time the upstream releases a new compiler version.

File: craft_parts/plugins/rust_plugin.py
Comment (cmatsuoka) [inline craft_parts/plugins/rust_plugin.py:199]: This would look nicer dedented.

Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Addressed.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Addressed

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Rust has a slightly different triplet system than us (and most compilers), so I think a different translation table may need to be used.

Comment (lengau) [reply craft_parts/plugins/rust_plugin.py]: Good point — there's also `self._part_info.target_arch` that'll provide the debian architecture name (`deb` in each dict of the mapping) if you'd prefer that to be the source. My naïve thought was to simply replace the first `-` in the triplet with `unknown`, but thinking about it further that neither covers armhf nor is any more reliable than just making a new mapping table.

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [reply craft_parts/plugins/rust_plugin.py:171]: Ahh, you're right — I missed the `'`s. Sorry about that!

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [reply craft_parts/plugins/rust_plugin.py]: That's a good point. In that case we probably do want to run `rustup update` since I would guess it would be less surprising to get a new version of rust than to come back to a project you haven't touched in a month and get an old version of rust.

Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: > That's a good point. In that case we probably do want to run `rustup update` since I would guess it would be less surprising to get a new version of rust than to come back to a project you haven't touched in a month and get an old version of rust.

For Rust, if you are using the `stable` channel (which is the default), it is improbable that the newer compiler would break your project. Even if it does, the Rust compiler usually will provide an option to fix your code automatically.

File: craft_parts/plugins/rust_plugin.py
Comment (syu-w) [inline craft_parts/plugins/rust_plugin.py:72]: Not only Snapcraft

File: craft_parts/plugins/rust_plugin.py
Comment (syu-w) [reply craft_parts/plugins/rust_plugin.py:131]: I would like to allow user to choice use system rust or install from rustup with/without pinned rust versions.

File: craft_parts/plugins/rust_plugin.py
Comment (syu-w) [reply craft_parts/plugins/rust_plugin.py:150]: Not sure if we want auto `cargo update`. We could add an option for that.

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [reply craft_parts/plugins/rust_plugin.py:131]: I like the use of rustup (and have dealt with [exactly the lagging rust versions issue](https://github.com/canonical/charmcraft/blob/main/snap/snapcraft.yaml#L74-L94) in charmcraft — though there `focal-updates` getting a new enough rust ended up resolving it). I've also dealt with the confusion of my core22 snap not having rust even when I used the plugin, so I'm all on board with that!

The backwards compatibility issue is within a single `coreX` base rather than between bases. While I personally prefer the core18 behaviour, we can't break existing `core22` snaps (or charms/rocks that use the plugin).

The two major ways people currently use to get rust on core22 are to install it in a `rust-deps` part or to add `cargo` to `build-packages` (as I did with charmcraft, and also like [sccache](https://github.com/mozilla/sccache/blob/main/snap/snapcraft.yaml#L40) and [alacritty](https://github.com/snapcrafters/alacritty/blob/1f478744ea69f1095c5399e65db2888b9d598ebe/snap/snapcraft.yaml#L75), and we need to ensure that neither of these break existing projects.

Perhaps the following behaviour will be more palatable:

1. If a `rust-deps` part dependency is defined, assume that installs rust and error if there's rustup configuration here.
  a) if there is a `rust-deps` part dependency, maybe log the fact that it's overriding the installation so users have a friendly way to know that they need to rename it
2. If `rust-channel` is undefined, check for `rustc` and `cargo` during the pull step. If they exist **and** `rustup` does not, no-op. (This should work since `build-packages` are installed before the pull step).
3. If we don't have `cargo` or `rustc`, do the `rustup` download/update logic.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:150]: I think that was a typo, they probably meant `rustup update`.

Comment (lengau) [reply craft_parts/plugins/rust_plugin.py:150]: You're right — I meant to say `cargo fetch`

Comment (lengau) [reply craft_parts/plugins/rust_plugin.py:150]: (Probably with an appropriate `--target`)

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [reply craft_parts/plugins/rust_plugin.py]: Exactly — and if you're using the `nightly` channel you can (theoretically at least) expect your code to break randomly in CI unless you've set up a `rust-toolchain.toml`.

But in a case like [the alacritty snap](https://github.com/snapcrafters/alacritty/blob/1f478744ea69f1095c5399e65db2888b9d598ebe/snap/snapcraft.yaml), if you have `rust-channel: stable` update the `source-tag`, you don't want it to tell you that rust is out of date and to have to `snapcraft clean` to get it to work — instead, re-pulling should include updating rust.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Fixed

**Comment by lengau** [comment]
A rebase should now fix the lint issues BTW :smile:

File: craft_parts/plugins/rust_plugin.py
Comment (sergiusens) [inline craft_parts/plugins/rust_plugin.py:45]: this is the part, as in
```yaml
parts:
  p1:
    plugin: rust
    source: .
    after: [rust-deps]
  rust-deps:
    plugin: nil
    override-build: |
      rustup...
```

File: craft_parts/plugins/rust_plugin.py
Comment (sergiusens) [inline craft_parts/plugins/rust_plugin.py:57]:
```yaml
after = values.get("after")
if value and "rust-deps" in after
```

File: craft_parts/plugins/rust_plugin.py
Comment (sergiusens) [inline craft_parts/plugins/rust_plugin.py:170]: also check if self._part.dependencies has rust-deps

Comment (cmatsuoka) [reply craft_parts/plugins/rust_plugin.py:193]: `self._part_info` is available to plugins (instead of `self_part`) and it currently doesn't provide dependencies. The information is currently passed only to environment validators. @liushuyu may have an alternative solution for this, otherwise we can add dependency information to `PartInfo`.

Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:193]: I have pushed an alternative solution yesterday:

https://github.com/canonical/craft-parts/pull/530/files#diff-f7845e9499f862b5d9c9c6140de217cea696e6c67f6dfb03cfe1bdf421861486R88-R99

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: This is not possible. `after` is not passed to this part.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Done.

File: craft_parts/plugins/rust_plugin.py
Comment (cmatsuoka) [inline craft_parts/plugins/rust_plugin.py:200]: Is "none" a valid channel value or something used for debug? It's not mentioned in the docstring above.

File: craft_parts/plugins/rust_plugin.py
Comment (cmatsuoka) [reply craft_parts/plugins/rust_plugin.py:195]: It seems that current solution detects if both the channel and a rust-deps part are available, however if only the rust-deps part is used, rustup is still invoked? `self._check_system_rust()` would fail at this point because the rust-deps part is not staged yet.

Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py:196]: > It seems that current solution detects if both the channel and a rust-deps part are available, however if only the rust-deps part is used, rustup is still invoked, `self._check_system_rust()` would fail at this point because the rust-deps part is not staged yet.

Addressed.
Now it will automatically set `"none"` in `validate_environment` if the user has `rust-deps` defined.

**Review by cmatsuoka — APPROVED** [review]
Thank you, I think we now cover all the requirements. @sergiusens, @lengau, @mr-cal, @syu-w, is there anything else you want to check?

**Comment by liushuyu** [comment]
CI errors do not seem related to this PR.

File: craft_parts/plugins/rust_plugin.py
Comment (lengau) [inline craft_parts/plugins/rust_plugin.py:166]: This is a real nitpick, but I think it's more readable to do

```suggestion
            rust_version = subprocess.check_output(["rustc", "--version"], text=True)
            cargo_version = subprocess.check_output(["cargo", "--version"], text=True)
            return "rustc" in rust_version and "cargo" in cargo_version
```

**Review by lengau — APPROVED** [review]
This is great, thank you! With some slight adjustments to charmcraft this will also solve most of our rust-dependencies issues in charms.

File: craft_parts/plugins/rust_plugin.py
Comment (liushuyu) [reply craft_parts/plugins/rust_plugin.py]: Done

**Comment by sergiusens** [comment]
the failing test might be obsolete now and chisel or the chisel DB might now be doing the right thing...
cc @cjdcordeiro for confirmation

**Comment by cjdcordeiro** [comment]
> the failing test might be obsolete now and chisel or the chisel DB might now be doing the right thing... cc @cjdcordeiro for confirmation

If I'm reading it right, the tests are failing on the case where Chisel is `! supported`, and the exception is not being raised. Chisel hasn't changed in quite some time. 

The test seems to be failing for 20.04, which you have marked as `unsupported` in 

https://github.com/canonical/craft-parts/blob/56f25859850f1694ab1d043deaa2712c8b1317c4/tests/integration/lifecycle/test_chisel_lifecycle.py#L32

(it is now supported though! could that be the issue?)

**Comment by liushuyu** [comment]
> > (it is now supported though! could that be the issue?)
> 
> This must be the cause, I'll create a PR to update the test.

I will rebase this pull request should the new pull request fixes the issue.

**Comment by Saviq** [comment]
The documentation could use an update following this:

https://forum.snapcraft.io/t/the-rust-plugin/8588

**Pull request merged**
